### PR TITLE
[benchmarks] Add scalar loss as model output when training

### PR DESCRIPTION
**Pull request by benjaminglass1**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #155877
* __->__ #158074

Add a hook to benchmark model forward passes that calculates a scalar loss as the first output when training (and detaches all other outputs).  This is a requirement to use joint graph export (experimental), but can be done without loss of generality.

Additionally ensures that Dynamo traces through the loss calculation function (not done previously), which reduces the number of graph breaks in models when training.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Comment by desertfire** [comment]
@anijain2305 , can you help to review this dynamo change?

**Comment by benjaminglass1** [comment]
@anijain2305 I ended up needing to add one new model XFAIL to this one, for `nvidia_deeprecommender`. I've spent several days trying to understand the failure, but effectively adding the loss return hook to that one model somehow changes the output of the forward pass? The failure is incredibly confusing, and I've been unable to determine any viable mechanism for it. Since this works for all other models in the training set, it seems reasonable to add a XFAIL for this one.

**Comment by benjaminglass1** [comment]
ROCm test failure appears unrelated to this PR.

**Comment by benjaminglass1** [comment]
Closing this, as we've decided not to pursue the AOTInductor training work at this time.
